src/math/polysat/viable.cpp

```diff
@@ -742,72 +778,183 @@ namespace {
+            if (fixed[i] == l_undef) {
+                lbool current = bound.get_bit(i) ? l_true : l_false;
+                if (carry) {
+                    if (FORWARD) {
+                        if (current == l_false) {
+                            new_bound[i] = l_true;
+                            carry = false;
+                        }
+                        else
+                            new_bound[i] = l_false;
+                    }
+                    else {
+                        if (current == l_true) {
+                            new_bound[i] = l_false;
+                            carry = false;
+                        }
+                        else
+                            new_bound[i] = l_true;
+                    }
+                }
+                else
+                    new_bound[i] = current;
+            }
+            else {
+                new_bound[i] = fixed[i];
+                if (carry)
+                    add_justification(i); // Again, we need this justification; if carry is false we don't need it
+            }
+        }
+        SASSERT(!src.empty());
+        if (carry) {
+            // We covered everything
+            /*if (FORWARD)
+                return rational::power_of_two(k);
+            else*/
+                return rational::zero();
+        }
+
+        // TODO: Directly convert new_bound in rational?
+        rational ret = rational::zero();
+        for (unsigned i = new_bound.size(); i > 0; i--) {
+            ret *= 2;
+            SASSERT(new_bound[i - 1] != l_undef);
+            ret += new_bound[i - 1] == l_true ? 1 : 0;
+        }
+        if (!FORWARD)
+            return ret + 1;
+        return ret;
+    }
+
+    // returns true iff no conflict was encountered
+    bool viable::collect_bit_information(pvar v, bool add_conflict, svector<lbool>& fixed, vector<ptr_vector<entry>>& justifications) {
+
+        pdd p = s.var(v);
+        // maybe pass them as arguments rather than having them as fields...
+        fixed.clear();
+        justifications.clear();
+        fixed.resize(p.power_of_2(), l_undef);
+        justifications.resize(p.power_of_2(), ptr_vector<entry>());
+
         auto* e = m_equal_lin[v];
         auto* first = e;
         if (!e)
             return true;
-        
-        pdd p = s.var(v);
+
         clause_builder builder(s, "bit check");
-        svector<lbool> fixed(p.power_of_2(), l_undef);
-        vector<ptr_vector<entry>> justifications(p.power_of_2(), ptr_vector<entry>());
         vector<std::pair<entry*, trailing_bits>> postponed;
-        
+
         auto add_entry = [&builder](entry* e) {
-            for (const auto& sc : e->side_cond) {
+            for (const auto& sc : e->side_cond)
                 builder.insert_eval(~sc);
-                LOG("Side cond: " << sc);
-            }
-            builder.insert_eval(~e->src);
-            LOG("Adding to core: " << e->src);
+            SASSERT(e->src.size() == 1);
```

Comment:
The assert holds because?
Generally tricky with these deep assumptions

Reply:
I dropped it. Left over from a previous state